Lasso Regression is a popular type of regularized linear regression that includes an L1 penalty. This has the effect of shrinking the coefficients for those input variables that do not contribute much to the prediction task. This penalty allows some coefficient values to go to the value of zero, allowing input variables to be effectively removed from the model, providing a type of automatic feature selection.

A popular penalty is to penalize a model based on the sum of the absolute coefficient values. This is called the L1 penalty. An L1 penalty minimizes the size of all coefficients and allows some coefficients to be minimized to the value zero, which removes the predictor from the model.

An L1 penalty minimizes the size of all coefficients and allows any coefficient to go to the value of zero, effectively removing input features from the model.

We can evaluate the Lasso Regression model on the housing dataset using repeated 10-fold cross-validation and report the average mean absolute error (MAE) on the dataset.

Ridge Regression is a popular type of regularized linear regression that includes an L2 penalty. This has the effect of shrinking the coefficients for those input variables that do not contribute much to the prediction task.

An L2 penalty minimizes the size of all coefficients, although it prevents any coefficients from being removed from the model by allowing their value to become zero.

# Example 

In [1]:
import numpy as np
import pandas as pd


In [2]:
# read dataset

dataset = pd.read_csv('Melbourne_housing_FULL.csv')
dataset.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Price          27247 non-null  float64
 5   Method         34857 non-null  object 
 6   SellerG        34857 non-null  object 
 7   Date           34857 non-null  object 
 8   Distance       34856 non-null  float64
 9   Postcode       34856 non-null  float64
 10  Bedroom2       26640 non-null  float64
 11  Bathroom       26631 non-null  float64
 12  Car            26129 non-null  float64
 13  Landsize       23047 non-null  float64
 14  BuildingArea   13742 non-null  float64
 15  YearBuilt      15551 non-null  float64
 16  CouncilArea    34854 non-null  object 
 17  Lattitude      26881 non-null  float64
 18  Longti

In [3]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount','Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]


In [5]:
dataset.head()

Suburb  Rooms Type Method SellerG             Regionname  \
0  Abbotsford      2    h     SS  Jellis  Northern Metropolitan   
1  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
2  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
3  Abbotsford      3    u     VB  Rounds  Northern Metropolitan   
4  Abbotsford      3    h     SP  Biggin  Northern Metropolitan   

   Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
3         4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
4         4019.0       2.5  Yarra City Council       3.0       2.0  0.0   

   Landsize  BuildingArea      Price  
0     126.0           NaN        NaN  
1     202.0           NaN  1480000.0  
2     156.0          79.0  1035000.0  
3       0.0           NaN        NaN  
4     134.0         150.0  1465000.0

In [4]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [5]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

In [6]:
dataset.isna().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Regionname          3
Propertycount       0
Distance            0
CouncilArea         3
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
Price            7610
dtype: int64

In [9]:
dataset.shape

(34857, 15)

In [7]:
dataset.dropna(inplace=True)

In [8]:
dataset.shape

(27244, 15)

In [12]:
dataset.isna().sum()

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Regionname       0
Propertycount    0
Distance         0
CouncilArea      0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
Price            0
dtype: int64

In [14]:
dataset.dtypes

Suburb            object
Rooms              int64
Type              object
Method            object
SellerG           object
Regionname        object
Propertycount    float64
Distance         float64
CouncilArea       object
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
Price            float64
dtype: object

In [9]:
dataset = pd.get_dummies(dataset, drop_first=True)

In [14]:
dataset.head(10)

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car    Landsize  \
1       2         4019.0       2.5       2.0       1.0  1.0  202.000000   
2       2         4019.0       2.5       2.0       1.0  0.0  156.000000   
4       3         4019.0       2.5       3.0       2.0  0.0  134.000000   
5       3         4019.0       2.5       3.0       2.0  1.0   94.000000   
6       4         4019.0       2.5       3.0       1.0  2.0  120.000000   
10      2         4019.0       2.5       2.0       1.0  0.0  181.000000   
11      3         4019.0       2.5       4.0       2.0  0.0  245.000000   
14      2         4019.0       2.5       2.0       1.0  2.0  256.000000   
15      3         4019.0       2.5       0.0       0.0  0.0  593.598993   
16      2         4019.0       2.5       0.0       0.0  0.0  593.598993   

    BuildingArea      Price  Suburb_Aberfeldie  ...  \
1       160.2564  1480000.0                  0  ...   
2        79.0000  1035000.0                  0  ...   
4       150.0000  1465000.0                  0  ...   
5       160.2564   850000.0                  0  ...   
6       142.0000  1600000.0                  0  ...   
10      160.2564   941000.0                  0  ...   
11      210.0000  1876000.0                  0  ...   
14      107.0000  1636000.0                  0  ...   
15      160.2564  1000000.0                  0  ...   
16      160.2564   745000.0                  0  ...   

    CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                     0                                  0   
2                                     0                                  0   
4                                     0                                  0   
5                                     0                                  0   
6                                     0                                  0   
10                                    0                                  0   
11                                    0                                  0   
14                                    0                                  0   
15                                    0                                  0   
16                                    0                                  0   

    CouncilArea_Nillumbik Shire Council  \
1                                     0   
2                                     0   
4                                     0   
5                                     0   
6                                     0   
10                                    0   
11                                    0   
14                                    0   
15                                    0   
16                                    0   

    CouncilArea_Port Phillip City Council  \
1                                       0   
2                                       0   
4                                       0   
5                                       0   
6                                       0   
10                                      0   
11                                      0   
14                                      0   
15                                      0   
16                                      0   

    CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                      0                                    0   
2                                      0                                    0   
4                                      0                                    0   
5                                      0                                    0   
6                                      0                                    0   
10                                     0                                    0   
11                                     0                                    0   
14                                     0                                    0   
15            

In [17]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27244 entries, 1 to 34856
Columns: 745 entries, Rooms to CouncilArea_Yarra Ranges Shire Council
dtypes: float64(8), int64(1), uint8(736)
memory usage: 21.2 MB


In [10]:
#Let's bifurcate our dataset into train and test dataset

X = dataset.drop('Price', axis=1)
y = dataset['Price']

In [11]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.3, random_state=2)

In [12]:
#Let's train our Linear Regression Model on training dataset and check the accuracy on test set

from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_X, train_y)

In [13]:
reg.score(test_X, test_y)

0.13853683161510377

In [14]:
reg.score(train_X, train_y)

0.6827792395792723

Here training score is 68% but test score is 13.85% which is very low

Normal Regression is clearly overfitting the data, let's try other models

In [16]:
#Using Lasso (L1 Regularized) Regression Model

from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(train_X, train_y)

C:\Users\DELL PC\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1326631542478295.2, tolerance: 796258688308225.0
  positive)


Lasso(alpha=50, copy_X=True, fit_intercept=True, max_iter=100, normalize=False,
      positive=False, precompute=False, random_state=None, selection='cyclic',
      tol=0.1, warm_start=False)

In [17]:
lasso_reg.score(test_X, test_y)



0.6636111369404487

In [18]:
lasso_reg.score(train_X, train_y)

0.6766985624766824

In [23]:
#Using Ridge (L2 Regularized) Regression Model

from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(train_X, train_y)

Ridge(alpha=50, copy_X=True, fit_intercept=True, max_iter=100, normalize=False,
      random_state=None, solver='auto', tol=0.1)

In [24]:
ridge_reg.score(test_X, test_y)

0.6670848945194957

In [25]:
ridge_reg.score(train_X, train_y)

0.6622376739684328

We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too

## Example 2

In [19]:
#Import numerical libraries
import pandas as pd
import numpy as np

#Import graphical plotting libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

#Import Linear Regression Machine Learning Libraries
from sklearn import preprocessing
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import r2_score

In [20]:
data = pd.read_csv('car-mpg.csv')
data.head()

mpg  cyl   disp   hp    wt   acc  yr  origin  car_type  \
0  18.0    8  307.0  130  3504  12.0  70       1         0   
1  15.0    8  350.0  165  3693  11.5  70       1         0   
2  18.0    8  318.0  150  3436  11.0  70       1         0   
3  16.0    8  304.0  150  3433  12.0  70       1         0   
4  17.0    8  302.0  140  3449  10.5  70       1         0   

                    car_name  
0  chevrolet chevelle malibu  
1          buick skylark 320  
2         plymouth satellite  
3              amc rebel sst  
4                ford torino

In [21]:
data.shape

(398, 10)

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   mpg       398 non-null    float64
 1   cyl       398 non-null    int64  
 2   disp      398 non-null    float64
 3   hp        398 non-null    object 
 4   wt        398 non-null    int64  
 5   acc       398 non-null    float64
 6   yr        398 non-null    int64  
 7   origin    398 non-null    int64  
 8   car_type  398 non-null    int64  
 9   car_name  398 non-null    object 
dtypes: float64(3), int64(5), object(2)
memory usage: 31.2+ KB


### EDA

In [23]:
#Drop car name
#Replace origin into 1,2,3.. dont forget get_dummies
#Replace ? with nan
#Replace all nan with median

data = data.drop(['car_name'], axis = 1)
data['origin'] = data['origin'].replace({1: 'america', 2: 'europe', 3: 'asia'})
data = pd.get_dummies(data,columns = ['origin'])


In [24]:
data.head()

mpg  cyl   disp   hp    wt   acc  yr  car_type  origin_america  \
0  18.0    8  307.0  130  3504  12.0  70         0               1   
1  15.0    8  350.0  165  3693  11.5  70         0               1   
2  18.0    8  318.0  150  3436  11.0  70         0               1   
3  16.0    8  304.0  150  3433  12.0  70         0               1   
4  17.0    8  302.0  140  3449  10.5  70         0               1   

   origin_asia  origin_europe  
0            0              0  
1            0              0  
2            0              0  
3            0              0  
4            0              0

In [27]:
data = data.replace('?', np.nan)


In [28]:
data.isnull().sum()

mpg               0
cyl               0
disp              0
hp                6
wt                0
acc               0
yr                0
car_type          0
origin_america    0
origin_asia       0
origin_europe     0
dtype: int64

In [29]:
data = data.apply(lambda x: x.fillna(x.median()), axis = 0)

In [30]:
data.head()

mpg  cyl   disp   hp    wt   acc  yr  car_type  origin_america  \
0  18.0    8  307.0  130  3504  12.0  70         0               1   
1  15.0    8  350.0  165  3693  11.5  70         0               1   
2  18.0    8  318.0  150  3436  11.0  70         0               1   
3  16.0    8  304.0  150  3433  12.0  70         0               1   
4  17.0    8  302.0  140  3449  10.5  70         0               1   

   origin_asia  origin_europe  
0            0              0  
1            0              0  
2            0              0  
3            0              0  
4            0              0

In [31]:
data.isnull().sum()

mpg               0
cyl               0
disp              0
hp                0
wt                0
acc               0
yr                0
car_type          0
origin_america    0
origin_asia       0
origin_europe     0
dtype: int64

In [ ]:
# We have to predict the mpg column given the features.

Split the dataset

In [32]:
X = data.drop(['mpg'], axis = 1) # independent variable
y = data[['mpg']] #dependent variable

In [34]:
y

mpg
0    18.0
1    15.0
2    18.0
3    16.0
4    17.0
..    ...
393  27.0
394  44.0
395  32.0
396  28.0
397  31.0

[398 rows x 1 columns]

In [35]:
X.describe()

cyl        disp           wt         acc          yr  \
count  398.000000  398.000000   398.000000  398.000000  398.000000   
mean     5.454774  193.425879  2970.424623   15.568090   76.010050   
std      1.701004  104.269838   846.841774    2.757689    3.697627   
min      3.000000   68.000000  1613.000000    8.000000   70.000000   
25%      4.000000  104.250000  2223.750000   13.825000   73.000000   
50%      4.000000  148.500000  2803.500000   15.500000   76.000000   
75%      8.000000  262.000000  3608.000000   17.175000   79.000000   
max      8.000000  455.000000  5140.000000   24.800000   82.000000   

         car_type  origin_america  origin_asia  origin_europe  
count  398.000000      398.000000   398.000000     398.000000  
mean     0.530151        0.625628     0.198492       0.175879  
std      0.499718        0.484569     0.399367       0.381197  
min      0.000000        0.000000     0.000000       0.000000  
25%      0.000000        0.000000     0.000000       0.000000  
50%      1.000000        1.000000     0.000000       0.000000  
75%      1.000000        1.000000     0.000000       0.000000  
max      1.000000        1.000000     1.000000       1.000000

In [36]:
#Scaling the data

X_s = preprocessing.scale(X)
X_s = pd.DataFrame(X_s, columns = X.columns) #converting scaled data into dataframe

y_s = preprocessing.scale(y)
y_s = pd.DataFrame(y_s, columns = y.columns) #ideally train, test data should be in columns

In [37]:
X_s

cyl      disp        hp        wt       acc        yr  car_type  \
0    1.498191  1.090604  0.673118  0.630870 -1.295498 -1.627426 -1.062235   
1    1.498191  1.503514  1.589958  0.854333 -1.477038 -1.627426 -1.062235   
2    1.498191  1.196232  1.197027  0.550470 -1.658577 -1.627426 -1.062235   
3    1.498191  1.061796  1.197027  0.546923 -1.295498 -1.627426 -1.062235   
4    1.498191  1.042591  0.935072  0.565841 -1.840117 -1.627426 -1.062235   
..        ...       ...       ...       ...       ...       ...       ...   
393 -0.856321 -0.513026 -0.479482 -0.213324  0.011586  1.621983  0.941412   
394 -0.856321 -0.925936 -1.370127 -0.993671  3.279296  1.621983  0.941412   
395 -0.856321 -0.561039 -0.531873 -0.798585 -1.440730  1.621983  0.941412   
396 -0.856321 -0.705077 -0.662850 -0.408411  1.100822  1.621983  0.941412   
397 -0.856321 -0.714680 -0.584264 -0.296088  1.391285  1.621983  0.941412   

     origin_america  origin_asia  origin_europe  
0          0.773559    -0.497643      -0.461968  
1          0.773559    -0.497643      -0.461968  
2          0.773559    -0.497643      -0.461968  
3          0.773559    -0.497643      -0.461968  
4          0.773559    -0.497643      -0.461968  
..              ...          ...            ...  
393        0.773559    -0.497643      -0.461968  
394       -1.292726    -0.497643       2.164651  
395        0.773559    -0.497643      -0.461968  
396        0.773559    -0.497643      -0.461968  
397        0.773559    -0.497643      -0.461968  

[398 rows x 10 columns]

In [38]:
X_s.describe()

cyl          disp            hp            wt           acc  \
count  3.980000e+02  3.980000e+02  3.980000e+02  3.980000e+02  3.980000e+02   
mean  -5.171742e-16 -8.312725e-17  1.458911e-16 -9.902743e-17  8.982206e-17   
std    1.001259e+00  1.001259e+00  1.001259e+00  1.001259e+00  1.001259e+00   
min   -1.444949e+00 -1.204411e+00 -1.527300e+00 -1.604943e+00 -2.747814e+00   
25%   -8.563206e-01 -8.563178e-01 -7.414364e-01 -8.828266e-01 -6.328794e-01   
50%   -8.563206e-01 -4.314040e-01 -2.830161e-01 -1.973624e-01 -2.472221e-02   
75%    1.498191e+00  6.584879e-01  5.421404e-01  7.538337e-01  5.834349e-01   
max    1.498191e+00  2.511784e+00  3.292662e+00  2.565185e+00  3.351912e+00   

                 yr      car_type  origin_america   origin_asia  origin_europe  
count  3.980000e+02  3.980000e+02    3.980000e+02  3.980000e+02   3.980000e+02  
mean  -8.837152e-16  9.623793e-16   -2.343184e-17 -5.219164e-16  -7.057448e-17  
std    1.001259e+00  1.001259e+00    1.001259e+00  1.001259e+00   1.001259e+00  
min   -1.627426e+00 -1.062235e+00   -1.292726e+00 -4.976433e-01  -4.619682e-01  
25%   -8.150739e-01 -1.062235e+00   -1.292726e+00 -4.976433e-01  -4.619682e-01  
50%   -2.721449e-03  9.414117e-01    7.735590e-01 -4.976433e-01  -4.619682e-01  
75%    8.096310e-01  9.414117e-01    7.735590e-01 -4.976433e-01  -4.619682e-01  
max    1.621983e+00  9.414117e-01    7.735590e-01  2.009471e+00   2.164651e+00

In [11]:
y_s

mpg
0   -0.706439
1   -1.090751
2   -0.706439
3   -0.962647
4   -0.834543
..        ...
393  0.446497
394  2.624265
395  1.087017
396  0.574601
397  0.958913

[398 rows x 1 columns]

In [ ]:
Steps:
    1. Analyzed the data
    2. Read the data
    3. Checked the summary of data
    4. Shape of data
    5. Preprocesseing:
        1. Find null values
        2. REplace null values/absured values with suitable values
        3. Standardization/normalization
        4. Visulalizations
        5. Find Outliers
        6. Remove outliers
        
    6. Split dataset
    7. Model
    8. Find accuracy
    
    
    
    
    

In [39]:
#Split into train, test set

X_train, X_test, y_train,y_test = train_test_split(X_s, y_s, test_size = 0.30, random_state = 1)
X_train.shape

(278, 10)

In [40]:
#Fit simple linear model and find coefficients
regression_model = LinearRegression()
regression_model.fit(X_train, y_train)

for idx, col_name in enumerate(X_train.columns):
    print('The coefficient for {} is {}'.format(col_name, regression_model.coef_[0][idx]))
    
intercept = regression_model.intercept_[0]
print('The intercept is {}'.format(intercept))

The coefficient for cyl is 0.32102238569161073
The coefficient for disp is 0.32483430918483835
The coefficient for hp is -0.22916950059437613
The coefficient for wt is -0.7112101905072296
The coefficient for acc is 0.014713682764190827
The coefficient for yr is 0.3755811949510746
The coefficient for car_type is 0.3814769484233105
The coefficient for origin_america is -0.07472247547584147
The coefficient for origin_asia is 0.044515252035678216
The coefficient for origin_europe is 0.04834854953945388
The intercept is 0.01928411610363974


In [41]:
regression_model.score(X_train, y_train)

0.8343770256960538

In [42]:
regression_model.score(X_test, y_test)

0.8513421387780066

## Regularized Ridge Regression

In [43]:

#alpha factor here is lambda (penalty term) which helps to reduce the magnitude of coeff

ridge_model = Ridge(alpha = 0.3)
ridge_model.fit(X_train, y_train)

print('Ridge model coef: {}'.format(ridge_model.coef_))
#As the data has 10 columns hence 10 coefficients appear here   

Ridge model coef: [[ 0.31649043  0.31320707 -0.22876025 -0.70109447  0.01295851  0.37447352
   0.37725608 -0.07423624  0.04441039  0.04784031]]


## Regularized Lasso Regression

In [45]:
#alpha factor here is lambda (penalty term) which helps to reduce the magnitude of coeff

lasso_model = Lasso(alpha = 0.1)
lasso_model.fit(X_train, y_train)

print('Lasso model coef: {}'.format(lasso_model.coef_))
#As the data has 10 columns hence 10 coefficients appear here   

Lasso model coef: [-0.         -0.         -0.01690287 -0.51890013  0.          0.28138241
  0.1278489  -0.01642647  0.          0.        ]


Here we notice many coefficients are turned to 0 indicating drop of those dimensions from the model

## Score Comparison

In [46]:
#Simple Linear Model
print(regression_model.score(X_train, y_train))
print(regression_model.score(X_test, y_test))

print('*************************')
#Ridge
print(ridge_model.score(X_train, y_train))
print(ridge_model.score(X_test, y_test))

print('*************************')
#Lasso
print(lasso_model.score(X_train, y_train))
print(lasso_model.score(X_test, y_test))

0.8343770256960538
0.8513421387780066
*************************
0.8343617931312616
0.851888217160851
*************************
0.7938010766228453
0.8375229615977083
